community/modules/scripts/htcondor-install/files/autoscaler.py

```diff
@@ -180,47 +181,30 @@ def scale(self):
         else:
             self.zoneargs = {"zone": self.zone}
 
-        # Get total number of jobs in the queue that includes number of jobs
-        # waiting as well as number of jobs already assigned to nodes
-        queue_length_req = (
-            'condor_q -allusers -totals -format "%d " Jobs -format "%d " Idle -format "%d " Held'
-        )
-        queue_length_resp = os.popen(queue_length_req).read().split()
+        # Count jobs that are running or idle ("potentially runnable"); do not
+        # count held jobs or jobs in transitioning states
+        # https://htcondor.readthedocs.io/en/latest/classad-attributes/job-classad-attributes.html#JobStatus
+        schedd = htcondor.Schedd()
+        job_attributes = [ "RequestCpus", "RequestMemory", "RequestGpus" ]
+        running_job_ads = schedd.query(constraint="JobStatus==2", projection=job_attributes)
+        idle_job_ads = schedd.query(constraint="JobStatus==1", projection=job_attributes)
 
-        if len(queue_length_resp) > 1:
-            queue = int(queue_length_resp[0])
-            idle_jobs = int(queue_length_resp[1])
-            on_hold_jobs = int(queue_length_resp[2])
-        else:
-            queue = 0
-            idle_jobs = 0
-            on_hold_jobs = 0
+        total_idle_request_cpus = sum(j["RequestCpus"] for j in idle_job_ads)
+        total_running_request_cpus = sum(j["RequestCpus"] for j in running_job_ads)
```

Comment:
Make "RequestCpus" a constant (all caps var), maybe other request strings as well although less important as they are not duplicated.

Reply:
A few lines above I link to HTCondor Job ClassAd attributes (for `JobStatus` and its defined values).

Would you like me to make that link more generic?

https://htcondor.readthedocs.io/en/latest/classad-attributes/job-classad-attributes.html#RequestCpus

Strictly speaking, the names of ClassAd attributes in HTCondor are defined to be case insensitive (though I can't find a link to that). What I can say is that every job gets a RequestCpus (and maybe RequestMemory, RequestGpus, RequestDisk?) with a default value.

So we can do what we want, but convention is CamelCase.

Reply:
I have added more descriptive comments and links that might help maintainers of this code who are less familiar with HTCondor administration.

Reply:
My request here was to the following between L187 and L188 (could also go at top of file)

```
REQUEST_CPU = 'RequestCpus'
```

And then use that variable in place of the string literals. This prevents typos and update errors.